README
		spamGuard 1.9 README
	---------------------------------------

Nearly all of today's mail system administrators face spam as their
first threat. Because of this, EnderUNIX team has written this small
application to automagically monitor malicious spammer activity in
your mail server logs.

spamGuard is written purely in C, to stop spammers hanging around.
The program supports nearly all mostly used MTAs; qmail (qSheff,
multilog, and splogger), sendmail, Postfix and Exim.

Rationale: 
----------

1. spamGuard is fired via crontab within fixed intervals. 
2. spamGuard scans the accumulated maillog and finds the amount of
mails users sent.
3. spamGuard loads its ignore list from its
   ignore file and badmailfiles files(badmailfrom in
   qmail and access for sendmail).
4. users whose mail count exceeds the configured threshold value(there
   are three different values: warning, block and paranoid) are treated 
   as spammers. They are added to mailers' badmail file (spammer 
   blacklist files).
5. Notification mails are sent to sysadmins about spammer activity,
   if present. 
6. (For sendmail version, a hash database of access
   file is created).

spamGuard can process a 85 MB log file in less then 10 seconds.

Tested on: FreeBSD 4-STABLE Solaris 2.7-Sparc
Solaris 9-Sparc Slackware Linux 8.0 RedHat 9, Digital UNIX, HP-UX

If your OS is DigitalUnix, please read README.DigitalUnix.

For questions, there is a mailing list available: To subscribe send
mail to: spamguard-subscribe@yahoogroups.com


EnderUNIX SDT @ Istanbul/Turkey bug-report at enderunix dot org -
http://www.enderunix.org

Tue May 08 21:08:08 EEST 2008


CHANGELOG

* Development version adds/changes the following:

* Thu Feb 21 23:37:12 EEST 2008
  Version 1.9 adds the following:
  - Exim support added. Mail adresses of spammers, found in exim main log file,
    according to bcnt, pcnt are added to badmail file.

 
* Wed May 23 22:14:34 EEST 2007
  Version 1.8 adds the following:
  - qSheff support added. IP adresses of spammers, found in qsheff.log, who send mails 
    which has same subject more than pcnt or bcnt value will be added to badmail file.
  - If subject filtering is enabled, subjects of spams, found in qsheff.log, will be added 
    to qsheff.rules file. To enabling this feature, make enable_subj_filt value 1.
  - If trimming subject string is enabled, subjects of spams, found in qsheff.log, will be added to
    qsheff.rules file after subjects were trimmed according to your trimming strings, such as [!! SPAM],
    [?? Probable Spam], [!!! VIRUS]. If the subject is "[!! SPAM] Mortgage", it will be trimmed as "Mortgage",
    and then it will be added to qsheff.rules file. To enabling this feature, make enable_subj_filt value 2.


* Fri May  6 21:13:48 EEST 2005
  - the email address of spammer were added to Subject.
  - New conf file  (spam-high.txt) added. the addresses in this file  will be added in badmail file.
    when they sent mail more than pcnt value.
  - the addresses are in ignore file never be added to badmailfrom file even if when  they sent mail
    more then pcnt. use spam-high.txt file to block.
  - spamguard configuration files were moved from  /usr/local/etc/ to  /usr/local/etc/spamguard directory.


* Tue Jul 21 09:38:00 EEST 2004
  Version 1.7 adds/changes the following:
  - Modifying the Makefile.am to allow for installing a sysconfdir.
    Sameh Attia <sameh.attia at tedata.net>
  - Added FAQ file.

* Tue May  28 18:03:21 EEST 2004
  Version 1.7-BETA adds/changes the following:
  - stat bug has been fixed.

  - #@[] address is ignored for qmail.

  - paranoid mode checks whether this address is in the 
    badmailfrom file already.

  - Strip '<' and '>' characters in e-mail addresses. (rmspace).

  - Lots of security fixes

  - parser module re-written. many security fixes
  
  - autogen.sh re-written to work with FreeBSD


* Mon Oct  7 10:58:20 EEST 2002
  Version 1.6 adds/changes the following:
  - The Parser module is now Isoqlog's parser.
 
  - Easy to use config file, spamguard.conf.
    Now the configuration parameters can be
    set at run time.

  - Removed MTA subdirs. 
    Installation has become very easy.
    

* Fri Mar 15 06:20:12 EET 2002
  Version 1.5 adds/changes the following:
  - Runs/tested on Solaris, Free/Net/Open BSD, Digital Unix,
    Linux.
  - Bug in ignore file processing code is fixed.
    If you are using version 1.4, we strongly suggest 
    that you upgrade to version 1.5.
    Because of this, one of the hash tables has been
    turned into linked list (ignore list).

  - New Feature:
    Now, there are three different threshold values,
    which are treated differently by spamGuard:
    warning, block and paranoid.

  - Some code refinement.


* Thu Feb 21 02:13:46 EET 2002
  Version 1.4 adds/changes the following:
  - Because of performance issues, 
    linked lists are replaced with hash tables. 

  - Postfix support.

  - Some more memory leaks are cleaned.


* Sun Feb 10 16:19:07 EET 2002
  Version 1.3 adds/changes the following:
  - Generic functions have been seperated from
    the MTA-dependant code base.

  - Qmail with syslog support.

  - Sendmail support.

  - Added code necessary to save and restore
    the previous position on maillog file, so that
    you won't need to rotate maillog.


* Tue Feb  5 00:43:45 EET 2002
  Version 1.2 adds/changes the following:
  - Wildcard support in ignore file

  - Fixed some memory leaks.
    (Thanks to Geoff Mitchell)

  - SPAMNUMBER is no longer hard-coded into binary.
    It's now taken from stdin, so you can experiment
    freely with your SPAMNUMBERs.


* Tue Feb  2 02:30:40 EET 2002
  Maintenance Release: 1.1

* Thu Jan 31 03:32:10 EET 2002
  First Public Release



loadconfig.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <ctype.h>

#include "loadconfig.h"

extern int errno;
extern int w;
extern int b;
extern int p;

void
loadconfig(const char *cfgfile)
{
	FILE *fd = NULL;
	char buf[BUFSIZE];
	char keyword[KEYSIZE];
	char value[VALSIZE];
	char *cp1, *cp2;
	char *variables[] = {
		"Invalid",
		"logtype",
		"logfile",
		"ignorefile",
		"badmailfile",
		"hostname",
		"sysadmin",
		"statfile",
		"mail_command",
		"makemap_command",
		"wcnt",
		"bcnt",
		"pcnt",
		"highfile",
		"enable_subj_filt",
		"qsheff_rules_file",
		"trim_subj_str"
	};

	int i = 0, j = 0, key = 0, line = 0, lenbuf = 0, keyword_nums = sizeof(variables)/sizeof(char *);

	if ((fd = fopen(cfgfile, "r")) == NULL) {
		fprintf(stderr, "loadconfig: cannot open spamguard configuration file %s, exiting...\n", cfgfile);
		exit(-1);
	}
	
	while ((fgets(buf, BUFSIZE, fd)) != NULL) {
		line++;
		if (buf[0] == '#') 
			continue;
		if ((lenbuf = strlen(buf)) <= 1)
			continue;
		cp1 = buf;
		cp2 = keyword;
		while (isspace((int)*cp1) && ((cp1 - buf) < lenbuf)) 
			cp1++;
		while (isgraph((int)*cp1) && (*cp1 != '=') && (j++ < KEYSIZE - 1) && ((cp1 - buf) < lenbuf)) 
			*cp2++ = *cp1++;
		*cp2 = '\0';
		j = 0;
		cp2 = value;
		while ((*cp1 != '\0') && (*cp1 !='\n') && (*cp1 !='=') && ((cp1 - buf) < lenbuf))
			cp1++;
		cp1++; 
		while (isspace((int)*cp1) && ((cp1 - buf) < lenbuf))
			cp1++; 
		if (*cp1 == '"') 
			cp1++;
		while ((*cp1 != '\0') && (*cp1 !='\n') && (*cp1 !='"') && (j++ < VALSIZE - 1) && ((cp1 - buf) < lenbuf))
			*cp2++ = *cp1++;
		*cp2-- = '\0';
		j = 0;
		if (keyword[0] =='\0' || value[0] =='\0')
			continue;
		key = 0;
		for (i = 0; i < keyword_nums; i++) {
			if ((strncmp(keyword, variables[i], KEYSIZE)) == 0) {
				key = i;
				break;
			}
		}

		switch(key) {
			case 0:
				fprintf(stderr, "Illegal Keyword: %s\n", keyword);
				break;
			case 1:
				strncpy(logtype, value, VALSIZE);
				break;
			case 2:
				strncpy(logfile, value, VALSIZE);
				break;
			case 3:
				strncpy(ignorefile, value, VALSIZE);
				break;
			case 4:
				strncpy(badmailfile, value, VALSIZE);
				break;
			case 5:
				strncpy(hostname, value, VALSIZE);
				break;
			case 6:
				strncpy(sysadmin, value, VALSIZE);
				break;
			case 7:
				strncpy(statfile, value, VALSIZE);
				break;
			case 8:
				strncpy(mail_command, value, VALSIZE);
				break;
			case 9:
				strncpy(makemap_command, value, VALSIZE);
				break;
			case 10:
				wcnt = atoi(value);
				w = 1;
				break;
			case 11:
				bcnt = atoi(value);
				b = 1;
				break;
			case 12:
				pcnt = atoi(value);
				p = 1;
				break;
			case 13:
				pcnt = atoi(value);
				strncpy(highfile, value, VALSIZE);
				break;
                        case 14:
                                enable_subj_filt = atoi(value);
                                break;
                        case 15:
                                strncpy(qsheff_rules_file, value, VALSIZE);
                                break;
                        case 16:
                                strncpy(trim_subj_str, value, VALSIZE);
                                break;
		}
	}

	if (fclose(fd) != 0) {
		fprintf(stderr, "File: %s - Line: %d: %s.\n", __FILE__, __LINE__, strerror(errno));
		exit(-1);
	}
}

void
readconfig(const char *cfgfile)
{
	struct stat statbuf;

	loadconfig(cfgfile);

	printf("logtype: %s\n", logtype);
	printf("logfile: %s\n", logfile);
	printf("ignorefile: %s\n", ignorefile);
	printf("highfile: %s\n", highfile);
	printf("badmailfile: %s\n", badmailfile);
	printf("statfile: %s\n", statfile);
	printf("warning count: %d\n", wcnt);
	printf("block count: %d\n", bcnt);
	printf("paranoid count: %d\n", pcnt);
        printf("qsheff subject filtering: %d\n", enable_subj_filt);
	printf("qsheff rules file: %s\n", qsheff_rules_file);
        printf("trim subject string: %s\n", trim_subj_str);

	if ((enable_subj_filt != 0) && (enable_subj_filt != 1) && (enable_subj_filt) != 2) {
                printf("enable_subj_filt value must be 0, 1 or 2\n");
                exit(-1);
        }

	if (b == 0) {
		printf("You must set block count(bcnt) value in spamguard.conf\n");
		exit(-1);
	}

	if (w && b && (wcnt >= bcnt)) {
		printf("Error!: warning count value:%d must be smaller than block count value:%d\n", wcnt, bcnt);
		exit(-1);
	}
	
	if (w && (wcnt >= pcnt)) {
		printf("Error!: warning count value:%d must be smaller than paranoid count value:%d\n", wcnt, pcnt);
		exit(-1);
	}
	
	if (b && (bcnt >= pcnt)) {
		printf("Error!: block count value:%d must be smaller than paranoid count value:%d\n", bcnt, pcnt);
		exit(-1);
	}

	if ((strlen(hostname)) <= 0 )
        	gethostname(hostname, VALSIZE);
	        printf("hostname: %s\n", hostname);

	if ((strcasecmp(logtype, "qmail") != 0) && (strcasecmp(logtype, "sendmail") != 0)
		&& (strcasecmp(logtype, "postfix") != 0)
		&& (strcasecmp(logtype, "qsheff") != 0) && (strcasecmp(logtype, "exim") != 0)) {
		printf("Invalid logtype: %s\naccepted logytpes are: "
		        "qmail, qsheff, exim, sendmail or postfix\n", logtype);
                exit(-1);
	}
	
	if ((stat(logfile, &statbuf)) == 0) {
		if((S_ISREG(statbuf.st_mode)) == 0) {
			fprintf(stderr, "You are using: %s log type "
					"logfile: %s  must be a regular file\n", logtype, logfile);
			exit(-1);
		}
	}
	
	else {
		fprintf(stderr, "You are using: %s log type "
				"logfile: %s file does not exist!\n", logtype, logfile);
		exit(-1);
	}

	if ((strlen(ignorefile)) <= 0) {
		printf("You must define a ignorefile\n");
		exit(-1);
	}
	
	if ((strlen(badmailfile)) <= 0) {
		printf("You must define a badmailfile\n");
		exit(-1);
	}
	
	if ((strlen(sysadmin)) <= 0) {
		printf("You must define a sysadmin email address\n");
		exit(-1);
	}

	if ((strlen(mail_command)) <= 0) {
		printf("You must set mail_command value\n");
		exit(-1);
	}
	
	if (((strcasecmp(logtype, "sendmail") == 0) || (strcasecmp(logtype, "postfix") == 0)) && (strlen(makemap_command) == 0)){
		printf("You are using logtype:%s, you must set makemap_command value\n", logtype);
		exit(-1);
	}

        if ((strlen(qsheff_rules_file)) <= 0) {
                printf("Wonna enabling qsheff subject filtering? Please specify a valid path for qsheff rules file\n");
                exit(-1);
	}

        if ((strlen(trim_subj_str) <= 0) && (enable_subj_filt == 2)) {
                printf("Wonna trimming subjects? Please type valid word(s) or change 'enable_subj_filt' value to 0 or 1\n");
                exit(-1);
        }
}
